File: src/main/java/com/example/reactive/chapter2/sse/TemperatureApplication.java
package com.example.reactive.chapter2.sse;

import java.util.concurrent.Executor;

import org.springframework.aop.interceptor.AsyncUncaughtExceptionHandler;
import org.springframework.aop.interceptor.SimpleAsyncUncaughtExceptionHandler;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.scheduling.annotation.AsyncConfigurer;
import org.springframework.scheduling.annotation.EnableAsync;
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;

@EnableAsync
@SpringBootApplication
public class TemperatureApplication implements AsyncConfigurer {

    public static void main(String[] args) {
        SpringApplication.run(TemperatureApplication.class, args);
    }

    @Override
    public Executor getAsyncExecutor() {

        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
        executor.setCorePoolSize(2);
        executor.setQueueCapacity(10);
        executor.setMaxPoolSize(10);
        executor.setThreadNamePrefix("sse-");
        executor.initialize();

        return executor;
    }

    @Override
    public AsyncUncaughtExceptionHandler getAsyncUncaughtExceptionHandler() {

        return new SimpleAsyncUncaughtExceptionHandler();
    }
}

File: src/main/kotlin/wmp/edu/reactive/sse/TemperatureApplication.kt
package wmp.edu.reactive.sse

import org.springframework.aop.interceptor.AsyncUncaughtExceptionHandler
import org.springframework.aop.interceptor.SimpleAsyncUncaughtExceptionHandler
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.scheduling.annotation.AsyncConfigurer
import org.springframework.scheduling.annotation.EnableAsync
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor
import java.util.concurrent.Executor

@EnableAsync
@SpringBootApplication
class TemperatureApplication : AsyncConfigurer {
    override fun getAsyncExecutor(): Executor? {
        val executorService = ThreadPoolTaskExecutor()
        executorService.corePoolSize = 5
        executorService.maxPoolSize = 10
        executorService.setQueueCapacity(10)
        executorService.setThreadNamePrefix("sse-")
        executorService.initialize()
        return executorService
    }

    override fun getAsyncUncaughtExceptionHandler(): AsyncUncaughtExceptionHandler? {
        return SimpleAsyncUncaughtExceptionHandler()
    }
}

fun main(args: Array<String>) {
    runApplication<TemperatureApplication>(*args)
}

